File: week02/example/ltc77_combinations.cpp
//
// Created by chenbingwei on 2022/4/29.
//
#include <bits/stdc++.h>

using namespace std;

class Solution {
public:
    vector<vector<int>> combine(int n, int k) {
        this->n = n;
        this->k = k;
        recur(1);
        return ans;

    }

private:
    vector<vector<int>> ans;
    vector<int> sub_combine;
    int n;
    int k;

    void recur(int i) {
        if (sub_combine.size() + (n - i + 1) < k) {
            return;
        }
        if (sub_combine.size() == k) {
            ans.push_back(sub_combine);
            return;
        }
        if (i > n) {
            return;
        }

        // 不选自己
        recur(i + 1);
        // 选自己
        sub_combine.push_back(i);
        recur(i + 1);
        sub_combine.pop_back();

    }

};

File: week02/example/ltc77_combinations.py
from typing import List


class Solution:
    def combine(self, n: int, k: int) -> List[List[int]]:
        tmp_combine = []
        ans = []

        def recur(num):
            if len(tmp_combine) > k or len(tmp_combine) + (n - num + 1) < k:
                return

            if num == n + 1:
                if len(tmp_combine) == k: ans.append(tmp_combine.copy())
                return
            # if len(tmp_combine) == k:
            #     ans.append(tmp_combine.copy())
            #     return
            # if num > n:
            #     return
            # next
            recur(num + 1)
            # save and next
            tmp_combine.append(num)
            recur(num + 1)
            tmp_combine.pop()

        recur(1)
        return ans


if __name__ == '__main__':
    out = Solution().combine(4, 2)
    print(out)
